content/blog/unsafe-string-interning-in-go.md

```diff
@@ -0,0 +1,345 @@
+)
+
+// We play unsafe games that violate Go's rules (and assume a non-moving
+// collector). So we quiet Go here.
+//go:nocheckptr
+
+// Get returns a pointer representing the comparable value cmpVal.
+//
+// The returned pointer will be the same for Get(v) and Get(v2)
+// if and only if v == v2, and can be used as a map key.
+func Get(cmpVal interface{}) *Value {
+	mu.Lock()
+	defer mu.Unlock()
+
+	var v *Value
+	addr, ok := val[cmpVal]; ok {
+		// Value is already interned in the pool. In case a finalizer runs in
+		// the near future, we must "resurrect" this value to bring it back
+		// to the active part of its lifecycle.
+		v = (*Value)((unsafe.Pointer)(addr))
+		v.resurrected = true
+	}
+	if v != nil {
+		return v
+	}
+
+	// Value must be created now and then interned. We store a uintptr to
+	// maintain a weak reference so that the value can eventually be garbage
+	// collected if it isn't used for a period of time.
+	v = &Value{cmpVal: cmpVal}
+	val[cmpVal] = uintptr(unsafe.Pointer(v))
+	runtime.SetFinalizer(v, finalize)
+
+	return v
+}
+
+// finalize is a Go finalizer for *Value.
+func finalize(v *Value) {
+	mu.Lock()
+	defer mu.Unlock()
+	if v.resurrected {
+		// We lost the race. Somebody resurrected it while we were about to
+		// finalize it. Try again next round.
+		v.resurrected = false
+		runtime.SetFinalizer(v, finalize)
+		return
+	}
+	delete(val, v.cmpVal)
+}
+```
+
+The code comments above discuss many of the ideas at work, but let's break
+down `Get` and `finalize` specifically to discuss how these interact with each
+other and the Go runtime.
+
+Starting with `Get`, we acquire the mutex around the map of interned values,
+which now stores the values as [weak references using
+`uintptr`](https://en.wikipedia.org/wiki/Weak_reference) rather than a concrete
+`*Value` type. This ensures that the values are not protected from garbage
+collection, and thus we must manually manage their lifecycles by using
+finalizers and the `v.resurrected` boolean field.
+
+First we check if the value is already present in the pool. If it is, we convert
+the weak reference back to a `*Value` by using `unsafe.Pointer` conversion, and
+mark the value as resurrected so that any finalizers which run concurrently will
+know that this value cannot be removed from the pool.
+
+```go
+	var v *Value
+	addr, ok := val[cmpVal]; ok {
+		// Value is already interned in the pool. In case a finalizer runs in
+		// the near future, we must "resurrect" this value to bring it back
+		// to the active part of its lifecycle.
+		v = (*Value)((unsafe.Pointer)(addr))
+		v.resurrected = true
+	}
+	if v != nil {
+		return v
+	}
+```
+
+If the value did not already exist in the pool, we can create it now and intern
+it there by adding it to the map of weak references. Note that the
+`v.resurrected` field defaults to `false`, meaning nobody has bothered to
+retrieve this value from the pool as of yet. We also install the finalizer for
```

Comment:
The wording obscures the subtlety of what's happening here. `resurrected` and the delaying logic in the finalizer is only necessary when we've materialized a *Value out of thin air in Get. IOW, we've violated the assumptions the runtime makes about pointers, so we have to do the `resurrected` trick to make sure we straighten things out.

In the path where we create *Value, that value's only existing reference in the program is a "strong" pointer that doesn't violate the runtime's rules. So, if nobody Get's it and we just drop this *Value, there's no trickery required to GC the value safely, because no rules have been violated yet.